README.md
# Chaos Toolkit Extension for SaltStack

[![Build Status](https://travis-ci.com/chaostoolkit-incubator/chaostoolkit-saltstack.svg?branch=master)](https://travis-ci.com/chaostoolkit-incubator/chaostoolkit-saltstack)
[![Python versions](https://img.shields.io/pypi/pyversions/chaostoolkit-saltstack.svg)](https://www.python.org/)

This project is a collection of [actions][] and [probes][], gathered as an
extension to the [Chaos Toolkit][chaostoolkit]. 

[actions]: http://chaostoolkit.org/reference/api/experiment/#action
[probes]: http://chaostoolkit.org/reference/api/experiment/#probe
[chaostoolkit]: http://chaostoolkit.org
[saltstack]: https://www.saltstack.com/

Generally [SaltStack][saltstack] runs minion agents in machine level, and minions are under management of one or more SaltMaster machines.
On the SaltMaster or via API, SaltMaster can run bulk commands or scripts or other actions on any connected minion.
The biggest advantage of SaltSatck is adapting cross cloud solutions on VM level chaos experiments.
If you are working on hybrid cloud and have network tunnelled, one SaltMaster could control all VMs across multiple clouds, e.g. Azure, AWS, etc.

## Install

This package requires Python 3.5+

To be used from your experiment, this package must be installed in the Python
environment where [chaostoolkit][] already lives.

```
$ pip install -U chaostoolkit-saltstack
```


## Usage

To use the probes and actions from this package, add the following to your
experiment file:

```json
{
    "type": "action",
    "name": "burn_cpu",
    "provider": {
        "type": "python",
        "module": "chaossaltstack.machines.actions",
        "func": "burn_cpu",
        "secrets": ["saltstack"],
        "arguments": {
            "parameters": {
                "execution_duration": "300"
            }
        }
    }
}
```

That's it!

Please explore the code to see existing probes and actions.



## Configuration

### Credentials
    0. Salt Master URL
        * SALTMASTER_HOST

    1. Useranme & Password.
       -d username='salt' -d password='abcd1234' -d eauth='pam'
       Then a token in obrained via <salt_url>/login

    2. Token
       A token directly, same with the backend of 1.

    3. Key, only from Salt Master
        * SALTMASTER_HOST: Salt Master API address

        You can authenticate with user / password via:
        * SALTMASTER_USER: the user name
        * SALTMASTER_PASSWORD: the password

        Or via a token:
        * SALTMASTER_TOKEN


There are two ways of doing this:

* you can either pass the name of the environment variables to the experiment definition as follows (recommended):

    ```json
    {
        "saltstack": {
            "SALTMASTER_HOST": {
                "type": "env",
                "key": "SALTMASTER_HOST"
            },
            "SALTMASTER_USER": {
                "type": "env",
                "key": "SALTMASTER_USER"
            },
            "SALTMASTER_PASSWORD": {
                "type": "env",
                "key": "SALTMASTER_PASSWORD"
            }
        }
    }
    ```

    ```json
    {
        "saltstack": {
            "SALTMASTER_HOST": {
                "type": "env",
                "key": "SALTMASTER_HOST"
            },
            "SALTMASTER_TOKEN": {
                "type": "env",
                "key": "SALTMASTER_TOKEN"
            }
        }
    }
    ```
    
* or you inject the secrets explicitly to the experiment definition:

    ```json
    {
        "saltstack": {
            "SALTMASTER_HOST": "https://172.10.20.666",
            "SALTMASTER_USER": "username",
            "SALTMASTER_PASSWORD": "password"
        }
    }
    ```

    ```json
    {
        "saltstack": {
            "SALTMASTER_HOST": "https://172.10.20.666",
            "SALTMASTER_TOKEN": "abcd1234abcd1234abcd1234"
        }
    }
    ```
    
    Additionally you may directly use if you are on the SaltMaster


### Putting it all together

Here is a full example:

```json
{
  "version": "1.0.0",
  "title": "...",
  "description": "...",
  "tags": [
    "azure",
    "kubernetes",
	"aks",
	"node"
  ],
  "configuration": {
    "saltstack": {
        "environment": "azure"
	}
  },
  "secrets": {
   "saltstack": {
       "SALTMASTER_HOST": "https://172.20.1.172:8000",
       "SALTMASTER_USER": "saltuser",
       "SALTMASTER_PASSWORD": "asfasfasdfa"
    }
   },
  "steady-state-hypothesis": {
    "title": "Services are all available and healthy",
    "probes": [
      {
        "type": "probe",
        "name": "check_minions_online",
        "provider": {
          "type": "python",
          "module": "chaossaltstack.machine.probes",
          "func": "is_minion_online",
          "arguments": {
              "instance_ids": [ "PABCDEFGS0016","PABCDEFGS0666" ]
          },
          "secrets": ["saltstack"]
        }
      }
    ]
  },
  "method": [
    {
        "type": "action",
        "name": "stress_cpu",
        "provider": {
            "type": "python",
          "module": "chaossaltstack.machine.probes",
            "module": "saltstack.machine.actions",
            "func": "stress_cpu",
            "arguments": {
                "execution_duration": "300",
                "instance_ids": [ "PABCDEFGS0016","PABCDEFGS0666" ]
            },
            "secrets": ["saltstack"]
        }
    }
  ],
  "rollbacks": [
    
  ]
}
```

## Contribute

If you wish to contribute more functions to this package, you are more than
welcome to do so. Please, fork this project, make your changes following the
usual [PEP 8][pep8] code style, sprinkling with tests and submit a PR for
review.

[pep8]: https://pycodestyle.readthedocs.io/en/latest/

The Chaos Toolkit projects require all contributors must sign a
[Developer Certificate of Origin][dco] on each commit they would like to merge
into the master branch of the repository. Please, make sure you can abide by
the rules of the DCO before submitting a PR.

[dco]: https://github.com/probot/dco#how-it-works

### Develop

If you wish to develop on this project, make sure to install the development
dependencies. But first, [create a virtual environment][venv] and then install
those dependencies.

[venv]: http://chaostoolkit.org/reference/usage/install/#create-a-virtual-environment

```console
$ pip install -r requirements-dev.txt -r requirements.txt 
```

Then, point your environment to this directory:

```console
$ pip install -e .
```

Now, you can edit the files and they will be automatically be seen by your
environment, even when running from the `chaos` command locally.

### Test

To run the tests for the project execute the following:

```
$ pytest
```


CHANGELOG.md
# Changelog

## [0.1.1][]

### Changed

- Bugfix: kill process now passes correct parameter
- Refactory: abstract core workflow into one common function to reduce duplicates

### Added

- Network experiments now accepts device name to adapt different os
- Add experiment action of kill processes by name
- Add experiment action of kill process by PID
- Add experiment probe to check process PID by name

### Changed

- Refactory: abstract core workflow into one common function to reduce duplicates

### Added

- Add experiment action of kill processes by name
- Add experiment action of kill process by PID
- Add experiment probe to check process PID by name

## [0.1.0][]

[0.1.0]: https://github.com/chaostoolkit-incubator/chaostoolkit-saltstack/tree/0.1.0

### Added

-   Initial release, experiments are only validated on Linux! Not Windows yet!
-   Add probe that check if a Salt Minion is online
-   Add probe that check if a Salt Minion is installed Linux tc for network experiments
-   Add experiment action of burn CPU to 100% usage
-   Add experiment action of burn I/O
-   Add experiment action of delay network
-   Add experiment action of network loss
-   Add experiment action of network corruption
-   Add experiment action of advanced network input
-   [Beta] Add experiment action of fill disk



tests/machine/test_machine_probes.py
from unittest.mock import MagicMock, patch

from chaossaltstack.machine.probes import is_minion_online, \
    is_iproute_tc_installed
from chaoslib.exceptions import FailedActivity
import pytest


THREE_INSTANCE = ["CLIENT1", "CLIENT2", "CLIENT3"]
ONLINE_MINION = ["CLIENT1"]
OFFLINE_MINION = ["CLIENT2"]
NO_MINION = ["CLIENT3"]
ONLINE_MINIONS = ["CLIENT1", "CLIENT4"]


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_minion_online_multiple_minions(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {"CLIENT1": True,"CLIENT2": False}
    client.run_cmd.return_value = cmd_return_value
    # # do
    res = is_minion_online(instance_ids=THREE_INSTANCE)
    # # assert
    assert res["CLIENT1"] == "Online"
    assert res["CLIENT2"] == "Offline"
    assert res["CLIENT3"] == "Not a Salt Minion"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_minion_online_offline_single(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {"CLIENT2": False}
    client.run_cmd.return_value = cmd_return_value
    # do
    res = is_minion_online(instance_ids=OFFLINE_MINION)

    assert res["CLIENT2"] == "Offline"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_minion_online_online_single(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {"CLIENT1": True}
    client.run_cmd.return_value = cmd_return_value

    res = is_minion_online(instance_ids=ONLINE_MINION)

    assert res["CLIENT1"] == "Online"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_minion_online_not_minion(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {"": ''}
    client.run_cmd.return_value = cmd_return_value

    res = is_minion_online(instance_ids=NO_MINION)

    assert res["CLIENT3"] == "Not a Salt Minion"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_minion_online_online_two(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {"CLIENT1": True, "CLIENT4": True}
    client.run_cmd.return_value = cmd_return_value

    res = is_minion_online(instance_ids=ONLINE_MINIONS)

    assert res["CLIENT4"] == "Online"
    assert res["CLIENT1"] == "Online"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_iproute_tc_installed_multiple_minions(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {'CLIENT1': 'Usage: tc xxxxx', 'CLIENT2': 'tc: command not found'}
    client.run_cmd.return_value = cmd_return_value
    # do
    res = is_iproute_tc_installed(instance_ids=THREE_INSTANCE)
    # assert
    assert res["CLIENT1"] == "Installed"
    assert res["CLIENT2"] == "Not Installed"
    assert res["CLIENT3"] == "Not a Salt Minion"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_iproute_tc_installed_single_not_installed(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {'CLIENT1': 'Usage: tc xxxxx'}
    client.run_cmd.return_value = cmd_return_value
    # do
    res = is_iproute_tc_installed(instance_ids=ONLINE_MINION)
    # assert
    assert res["CLIENT1"] == "Installed"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_iproute_tc_installed_single_installed(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {'CLIENT2': 'tc: command not found'}
    client.run_cmd.return_value = cmd_return_value
    # do
    res = is_iproute_tc_installed(instance_ids=OFFLINE_MINION)
    # assert
    assert res["CLIENT2"] == "Not Installed"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_iproute_tc_installed_two_installed(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {'CLIENT1': 'Usage: tc xxxxx', 'CLIENT4': 'Usage: tc xxxxx'}
    client.run_cmd.return_value = cmd_return_value
    # do
    res = is_iproute_tc_installed(instance_ids=ONLINE_MINIONS)
    # assert
    assert res["CLIENT1"] == "Installed"
    assert res["CLIENT4"] == "Installed"


@patch('chaossaltstack.machine.probes.saltstack_api_client', autospec=True)
def test_is_iproute_tc_installed_not_minion(init):
    # mock
    client = MagicMock()
    init.return_value = client
    cmd_return_value = {'': ''}
    client.run_cmd.return_value = cmd_return_value
    # do
    res = is_iproute_tc_installed(instance_ids=NO_MINION)
    # assert
    assert res["CLIENT3"] == "Not a Salt Minion"



ci.bash
#!/bin/bash
set -eo pipefail

function lint () {
    echo "Checking the code syntax"
    pylama chaossaltstack
}

function build () {
    echo "Building the chaostoolkit-saltstack package"
    python setup.py build
}

function run-test () {
    echo "Running the tests"
    python setup.py test
}

function release () {
    echo "Releasing the package"
    python setup.py release

    echo "Publishing to PyPI"
    pip install twine
    twine upload dist/* -u ${PYPI_USER_NAME} -p ${PYPI_PWD}
}

function main () {
    lint || return 1
    build || return 1
    run-test || return 1

    if [[ $TRAVIS_PYTHON_VERSION =~ ^3\.5+$ ]]; then
        if [[ $TRAVIS_TAG =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
            echo "Releasing tag $TRAVIS_TAG with Python $TRAVIS_PYTHON_VERSION"
            release || return 1
        fi
    fi
}

main "$@" || exit 1
exit 0


chaossaltstack/machine/scripts/network_advanced.sh
#!/bin/bash

#Mandatory parameters:
#param='delay 1000ms 500ms' or param='delay 1000ms 500ms' or param='corrupt 15%'
#duration=60

#Check if it is a normal setting
check=$(tc -s qdisc show dev $device |grep pfifo_fast)
ret=$?

#if not, recovery
if [ $ret -ne 0 ]; then
    tc qdisc del dev $device root 2>&1 >/dev/null
fi

#Do experiments
experiment=$(tc qdisc add dev $device root netem $param)
ret=$?
if [ $ret -eq 0 ]; then
    echo "experiment network_latency -> <$instance_id>: success"
else
    echo "experiment network_latency -> <$instance_id>: fail"
fi

#Sleep $duration
#Mandatory
sleep $duration

#recovery
tc -s qdisc show dev $device |grep pfifo_fast 2>&1 >/dev/null
ret1=$?
#if not, recovery
if [ $ret1 -ne 0 ]; then
    tc qdisc del dev $device root 2>&1 >/dev/null
fi

#Check if it is a normal setting
tc -s qdisc show dev $device |grep pfifo_fast 2>&1 >/dev/null
ret2=$?
if [ $ret2 -eq 0 ]; then
    echo "recover network_latency -> <$instance_id>: success"
else
    echo "recover network_latency -> <$instance_id>: fail"
fi

chaossaltstack/machine/actions.py
# -*- coding: utf-8 -*-
import os
import json
from time import sleep
from typing import List

from chaoslib.exceptions import FailedActivity
from chaoslib.types import Configuration, Secrets
from logzero import logger

from .. import saltstack_api_client
from ..types import SaltStackResponse
from .constants import OS_LINUX, OS_WINDOWS
from .constants import BURN_CPU, FILL_DISK, NETWORK_UTIL, \
    BURN_IO, KILLALL_PROCESSES, KILL_PROCESS


__all__ = ["burn_cpu", "fill_disk", "network_latency", "burn_io",
           "network_loss", "network_corruption", "network_advanced",
           "killall_processes", "kill_process"]


def burn_cpu(instance_ids: List[str] = None,
             execution_duration: str = "60",
             configuration: Configuration = None,
             secrets: Secrets = None) -> SaltStackResponse:
    """
    burn CPU up to 100% at random machines.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Duration of the stress test (in seconds) that generates high CPU usage.
        Defaults to 60 seconds.
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """

    logger.debug(
        "Start burn_cpu: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=BURN_CPU,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def fill_disk(instance_ids: List[str] = None,
              execution_duration: str = "120",
              size: str = "1000",
              configuration: Configuration = None,
              secrets: Secrets = None) -> SaltStackResponse:
    """
    For now do not have this scenario, fill the disk with random data.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Lifetime of the file created. Defaults to 120 seconds.
    size : str
        Size of the file created on the disk. Defaults to 1GB.
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """

    logger.debug(
        "Start fill_disk: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["execution_duration"] = execution_duration
    param["size"] = size

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=FILL_DISK,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def burn_io(instance_ids: List[str] = None,
            execution_duration: str = "60",
            configuration: Configuration = None,
            secrets: Secrets = None) -> SaltStackResponse:
    """
    Increases the Disk I/O operations per second of the virtual machine.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Lifetime of the file created. Defaults to 120 seconds.
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """

    logger.debug(
        "Start burn_io: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=BURN_IO,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def network_advanced(instance_ids: List[str] = None,
                     execution_duration: str = "60",
                     command: str = "",
                     device: str = "eth0",
                     configuration: Configuration = None,
                     secrets: Secrets = None) -> SaltStackResponse:
    """
    do a customized operations on the virtual machine via Linux - TC.
    For windows, no solution as for now.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Lifetime of the file created. Defaults to 60 seconds.
    command : str
        the tc command, e.g.  loss 15%
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """

    logger.debug(
        "Start network_advanced: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration
    param["param"] = command
    param["device"] = device

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=NETWORK_UTIL,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def network_loss(instance_ids: List[str] = None,
                 execution_duration: str = "60",
                 loss_ratio: str = "5%",
                 device: str = "eth0",
                 configuration: Configuration = None,
                 secrets: Secrets = None) -> SaltStackResponse:
    """
    do a network loss operations on the virtual machine via Linux - TC.
    For windows, no solution as for now.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Lifetime of the file created. Defaults to 60 seconds.
    loss_ratio : str:
        loss_ratio = "30%"
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """

    logger.debug(
        "Start network_advanced: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration
    param["param"] = "loss " + loss_ratio
    param["device"] = device
    
    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=NETWORK_UTIL,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def network_corruption(instance_ids: List[str] = None,
                       execution_duration: str = "60",
                       corruption_ratio: str = "5%",
                       device: str = "eth0",
                       configuration: Configuration = None,
                       secrets: Secrets = None) -> SaltStackResponse:
    """
    do a network loss operations on the virtual machine via Linux - TC.
    For windows, no solution as for now.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Lifetime of the file created. Defaults to 60 seconds.
    corruption_ratio : str:
        corruption_ratio = "30%"
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """

    logger.debug(
        "Start network_corruption: configuration='{}', "
        "instance_ids='{}'".format(configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration
    param["param"] = "corrupt " + corruption_ratio
    param["device"] = device


    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=NETWORK_UTIL,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def network_latency(instance_ids: List[str] = None,
                    execution_duration: str = "60",
                    delay: str = "1000ms",
                    variance: str = "500ms",
                    ratio: str = "",
                    device: str = "eth0",
                    configuration: Configuration = None,
                    secrets: Secrets = None) -> SaltStackResponse:
    """
    Increases the response time of the virtual machine.

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional
        Lifetime of the file created. Defaults to 120 seconds.
    delay : str
        Added delay in ms. Defaults to 1000ms.
    variance : str
        Variance of the delay in ms. Defaults to 500ms.
    ratio: str = "5%", optional
        the specific ratio of how many Variance of the delay in ms.
        Defaults to "".
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """
    logger.debug(
        "Start network_latency: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration
    param["param"] = "delay " + delay + " " + variance + " " + ratio
    param["device"] = device

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=NETWORK_UTIL,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def killall_processes(instance_ids: List[str] = None,
                      execution_duration: str = "60",
                      process_name: str = None,
                      configuration: Configuration = None,
                      signal: str = "",
                      secrets: Secrets = None) -> SaltStackResponse:
    """
    The killall utility kills processes selected by name
    refer to https://linux.die.net/man/1/killall

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional default to 1 second
        This is not technically not useful as the process usually is killed
        without and delay, however you can set more seconds here to let the
        thread wait for more time to extend your experiment execution in case
        you need to watch more on the observation metrics.
    process_name : str
        Name of the process to be killed
    signal : str , default to ""
        The signal of killall command, e.g. use -9 to force kill
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """
    logger.debug(
        "Start network_latency: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration
    param["process_name"] = process_name
    param["signal"] = signal

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=KILLALL_PROCESSES,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


def kill_process(instance_ids: List[str] = None,
                 execution_duration: str = "60",
                 process: str = None,
                 configuration: Configuration = None,
                 signal: str = "",
                 secrets: Secrets = None) -> SaltStackResponse:
    """
    kill -s [signal_as_below] [processname]
    HUP INT QUIT ILL TRAP ABRT EMT FPE KILL BUS SEGV SYS PIPE ALRM TERM URG
    STOP TSTP CONT CHLD TTIN TTOU IO XCPU XFSZ VTALRM PROF WINCH INFO USR1 USR2

    Parameters
    ----------
    instance_ids : List[str]
        Filter the virtual machines. If the filter is omitted all machines in
        the subscription will be selected as potential chaos candidates.
    execution_duration : str, optional default to 1 second
        This is not technically not useful as the process usually is killed
        without and delay, however you can set more seconds here to let the
        thread wait for more time to extend your experiment execution in case
        you need to watch more on the observation metrics.
    process : str
        pid or process that kill command accepts
    signal : str , default to ""
        The signal of kill command, use kill -l for help
    configuration : Configuration
        Chaostoolkit Configuration
    secrets : Secrets
        Chaostoolkit Secrets
    """
    logger.debug(
        "Start network_latency: configuration='{}', instance_ids='{}'".format(
            configuration, instance_ids))

    param = dict()
    param["duration"] = execution_duration
    param["process_name"] = process
    param["signal"] = signal

    return __default_salt_experiment__(instance_ids=instance_ids,
                                       execution_duration=execution_duration,
                                       param=param,
                                       experiment_type=KILL_PROCESS,
                                       configuration=configuration,
                                       secrets=secrets
                                       )


###############################################################################
# Private helper functions
###############################################################################
def __default_salt_experiment__(instance_ids: List[str] = None,
                                execution_duration: str = "60",
                                param: dict = None,
                                experiment_type: str = None,
                                configuration: Configuration = None,
                                secrets: Secrets = None
                                ) -> SaltStackResponse:
    response = dict()
    try:
        client = saltstack_api_client(secrets)
        machines = client.get_grains_get(instance_ids, 'kernel')

        jids = dict()

        if len(machines) <= 0:
            raise FailedActivity(
                "Cannot find any machines {}".format(instance_ids))

        for k, v in machines.items():
            name = k
            os_type = v
            param["instance_id"] = k
            script_content = __construct_script_content__(
                experiment_type, os_type, param)

            # Do async cmd and get jid
            logger.debug("{0} of machine: {1}".format(experiment_type, name))
            salt_method = 'cmd.run'
            jid = client.async_run_cmd(name, salt_method, script_content)
            jids[k] = jid
        logger.debug("SaltStack return jids:\n{}".format(json.dumps(jids)))
        # Wait the duration as well
        sleep(int(execution_duration))

        # Check result
        for k, v in jids.items():
            res = client.async_cmd_exit_success(v)[k]
            result = client.get_async_cmd_result(v)[k]
            if 'fail' in result:
                res = False
            response_item = "Machine {0} : {1} - Console: {2}".format(
                k, res, result)
            response[k] = response_item
        return response
    except Exception as x:
        raise FailedActivity(
            "failed issuing a execute of shell script via salt API " + str(x)
        )


def __construct_script_content__(action, os_type, parameters):

    if os_type == OS_WINDOWS:
        script_name = action+".ps1"
        # TODO in ps1
        cmd_param = '\n'.join(
            ['='.join([k, "'"+v+"'"]) for k, v in parameters.items()])
    elif os_type == OS_LINUX:
        script_name = action+".sh"
        cmd_param = '\n'.join(
            ['='.join([k, "'"+v+"'"]) for k, v in parameters.items()])
    else:
        raise FailedActivity(
            "Cannot find corresponding script for {} on OS: {}".format(
                action, os_type))

    with open(os.path.join(os.path.dirname(__file__),
                           "scripts", script_name)) as file:
        script_content = file.read()
    # merge duration
    script_content = cmd_param + "\n" + script_content
    return script_content
